### PR TITLE
Add a --output-type flag to generate different output artifacts

### PR DESCRIPTION
**Pull request by Itxaka**
Fixes https://github.com/kairos-io/kairos/issues/2171
Fixes https://github.com/kairos-io/kairos/issues/2202

adds a --output-type flag that allows to set the output type. Currently iso, artifact and container are supported and will generate the proper artifacts.

 - iso: Will generate an iso in the output dir
 - artifact: Will generate the UKI EFI files, configs and keys in the output dir
 - container: Will generate a container image witht he artifacts inside them (Valid for upgrade)

This changes the `--output` flag from a specific iso file to a dir (`--output-dir`), to be able to generate the output artifacts using that dir as a base. Name is now generated automatically based on the version obtained from the source.

Iso will now be kairos_VERSION.iso
Artifact/config names are kept the same as before
Container will be kairos_uki:$VERSION

### CODE REVIEW

File: pkg/utils/common.go [hunk 1]
```diff
@@ -58,3 +66,56 @@ func GetUkiCmdline() []string {
 	}
 
 }
+
+// LayerFromDir Converts a directory to a container layer by tarballing it
+func LayerFromDir(root string) (container.Layer, error) {
```
Comment (mauromorales) [inline pkg/utils/common.go:71]: fancy

**Review by mauromorales — APPROVED** [review]

**Comment by mudler** [comment]
I find the "artifact" output type confusing from a user standpoint, maybe let's just call it UKI?

File: pkg/action/build-uki.go [hunk 2]
```diff
@@ -49,47 +55,72 @@ func (b *BuildUKIAction) Run() error {
...
+		err = b.createISO(sourceDir)
+		b.logger.Infof("Done building %s at: %s", b.artifact, b.outputDir)
+	case string(constants.ContainerOutput):
+		// First create the files
+		err = b.createArtifact(sourceDir)
+		if err != nil {
+			return err
+		}
+		// Then build the image
+		kairosVersion, _ := findKairosVersion(sourceDir)
```
Comment (mudler) [inline pkg/action/build-uki.go:103]: shouldn't we fail if we don't find the version here? as it actually output files are named after it

**Comment by mudler** [comment]
Question: the systemd-boot configs are not generated in this step, do I read it right?

**Comment by mauromorales** [comment]
> I find the "artifact" output type confusing from a user standpoint, maybe let's just call it UKI?

I agree, but we are using artifact also in our docs for all other types of artifacts including container images and framework images.

what about `output-dir` and `output`? or `output-type` or more confusing?

File: pkg/action/build-uki.go (hunk 2 above)
Comment (Itxaka) [reply pkg/action/build-uki.go:103]: yes!

**Comment by Itxaka** [comment]
> Question: the systemd-boot configs are not generated in this step, do I read it right?

no you are not reading it rigth, it generates the full thing before generating the output, so by the time we go to the output, the files are already generated.
iso -> picks those files and puts them into an iso
artifact -> just copies the files into the output dir
container -> builds container from those files

**Comment by Itxaka** [comment]
> > I find the "artifact" output type confusing from a user standpoint, maybe let's just call it UKI?
> 
> I agree, but we are using artifact also in our docs for all other types of artifacts including container images and framework images.
> 
> what about `output-dir` and `output`? or `output-type` or more confusing?

yes, i hated the artifact name but could not decide on a proper naming for it. This output-dir/output-type sounds good to me!

File: pkg/constants/constants.go [hunk 3]
```diff
@@ -6,6 +6,16 @@ import (
 	"path/filepath"
 )
 
+type UkiOutput string
+
+const IsoOutput UkiOutput = "iso"
+const ContainerOutput UkiOutput = "container"
+const DefaultOutput UkiOutput = "artifact"
```
Comment (mudler) [inline pkg/constants/constants.go:13]: this is actually what I meant originally (woops)..

we call this const "DefaultOutput" but we refer to "artifact" which is quite generic, but aren't we just generating UKI files and its configs?  

would make sense to me to call this "UKI", so it's more self-explanatory (and you don't have to explain what kind of artifacts are you generating here everywhere)

Comment (Itxaka) [reply pkg/constants/constants.go:13]: ahh yeah, I gotchu.

So change this into `UKI` ?

**Comment by mudler** [comment]
one last thought here: maybe we should just build all of them at once? isn't easier also to document?

**Comment by Itxaka** [comment]
> one last thought here: maybe we should just build all of them at once? isn't easier also to document?

IMO too slow and uneeded. why would you get so many artifacts hen you only want one? is there a reason to get the artifacts in 3 different ways when you either want one or the other?

I mean, we could work out on using the flag multiple times to say what outputs you want and just iterate over them but Im, not sure its needed no?

File: pkg/constants/constants.go (hunk 3 above)
Comment (mudler) [reply pkg/constants/constants.go:13]: yep yep, easier to grasp for everyone, less to explain

Comment (mauromorales) [reply pkg/constants/constants.go]: ah yeah now I see what you meant 👍

File: Earthfile [hunk 4]
```diff
@@ -12,8 +12,9 @@ test:
     FROM golang:$GO_VERSION
     RUN apk add rsync gcc musl-dev docker jq
     WORKDIR /build
-    COPY . .
+    COPY go.mod go.sum . # This will make the go mod download able to be cached as long as it hasnt change
```
Comment (mauromorales) [inline Earthfile:15]: 👏

**Review by mauromorales — APPROVED** [review]

**Comment by Itxaka** [comment]
> > one last thought here: maybe we should just build all of them at once? isn't easier also to document?
> 
> IMO too slow and uneeded. why would you get so many artifacts hen you only want one? is there a reason to get the artifacts in 3 different ways when you either want one or the other?
> 
> I mean, we could work out on using the flag multiple times to say what outputs you want and just iterate over them but Im, not sure its needed no?

I dont know if that mean you agree with me on not doing it or you agree on specifying the flag multiple times :D

Let me know so Ic an either rework this to get the output flag multiple times and output more than one artifact or I can merge as is :D

File: pkg/constants/constants.go (hunk 3 above)
Comment (Itxaka) [reply pkg/constants/constants.go]: doneeee, no idea why github doesnt say its outdated....

**Comment by mudler** [comment]
> > > one last thought here: maybe we should just build all of them at once? isn't easier also to document?
> > 
> > 
> > IMO too slow and uneeded. why would you get so many artifacts hen you only want one? is there a reason to get the artifacts in 3 different ways when you either want one or the other?
> > I mean, we could work out on using the flag multiple times to say what outputs you want and just iterate over them but Im, not sure its needed no?
> 
> @mudler I dont know if that means you agree with me on not doing it or you agree on specifying the flag multiple times :D
> 
> Let me know so Ic an either rework this to get the output flag multiple times and output more than one artifact or I can merge as is

was just thinking out loud, sounds good!

**Review by mudler — APPROVED** [review]

**Pull request merged**